pkg/virtctl/set/set.go

```diff
@@ -0,0 +1,170 @@
+		Example: setUsage(),
+		Args:    templates.ExactArgs("set", 1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return run(cmd, args, clientConfig)
+		},
+	}
+
+	cmd.Flags().Uint32("cpu", 0, "Number of CPUs (sockets) to set (must be greater than 0)")
+	cmd.Flags().String("memory", "", "Memory size to set (e.g., 2Gi)")
+	cmd.Flags().Duration("timeout", 10*time.Second, "Timeout duration (e.g., 10s, 2m)")
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+	return cmd
+}
+
+func setUsage() string {
+	return `  # Set the number of CPUs (sockets) to 2 and memory to 1Gi for VirtualMachine 'myvm':
+  {{ProgramName}} set myvm --cpu=2 --memory=1Gi`
+}
+
+func run(cmd *cobra.Command, args []string, clientConfig clientcmd.ClientConfig) error {
+	vmName := args[0]
+	namespace, _, err := clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+
+	cpuSocketsCount, err := cmd.Flags().GetUint32("cpu")
+	if err != nil {
+		return err
+	}
+
+	guestMemorySizeStr, err := cmd.Flags().GetString("memory")
+	if err != nil {
+		return err
+	}
+
+	timeout, err := cmd.Flags().GetDuration("timeout")
+	if err != nil {
+		return err
+	}
+
+	var cpuSet, memorySet bool
+	var guestMemorySize resource.Quantity
+
+	if cmd.Flags().Changed("cpu") {
+		cpuSet = true
+		if cpuSocketsCount == 0 {
+			return fmt.Errorf("invalid CPU count: %d; must be greater than 0", cpuSocketsCount)
+		}
+	}
+
+	if cmd.Flags().Changed("memory") {
+		memorySet = true
+		guestMemorySize, err = resource.ParseQuantity(guestMemorySizeStr)
+		if err != nil {
+			return fmt.Errorf("invalid memory size: %s", guestMemorySizeStr)
+		}
+		if guestMemorySize.IsZero() || guestMemorySize.Sign() == -1 {
+			return fmt.Errorf("memory size must be greater than zero")
+		}
+	}
+
+	if !cpuSet && !memorySet {
+		return fmt.Errorf("at least one of --cpu or --memory must be set")
+	}
+
+	virtCli, err := kubecli.GetKubevirtClientFromClientConfig(clientConfig)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	vm, err := virtCli.VirtualMachine(namespace).Get(ctx, vmName, k8smetav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	messages, err := setCPUAndMemory(ctx, virtCli, vm, cpuSet, cpuSocketsCount, memorySet, guestMemorySize)
+	if err != nil {
+		return err
+	}
+
+	for _, message := range messages {
+		cmd.Printf("%s\n", message)
+	}
+	return nil
+}
+
+func setCPUAndMemory(ctx context.Context, virtCli kubecli.KubevirtClient, vm *v1.VirtualMachine, cpuSet bool, cpuSocketsCount uint32, memorySet bool, guestMemorySize resource.Quantity) ([]string, error) {
+	var messages []string
+	patchSet := patch.New()
+
+	if cpuSet {
+		cpuToPatch := vm.Spec.Template.Spec.Domain.CPU
+		if cpuToPatch == nil {
+			cpuToPatch = &v1.CPU{}
+		}
+		cpuToPatch.Sockets = cpuSocketsCount
+
+		patchSet.AddOption(patch.WithReplace("/spec/template/spec/domain/cpu", cpuToPatch))
```

Comment:
You never answered my question, did you try what happens if the running VM does not have a CPU specified? Does Replace really work?

Reply:
the replace operation works even if the vm doesn't have the specified `/spec/template/spec/domain/cpu`
it will simply add this field

Reply:
This contradicts the definition of the `replace` operation in the json patch spec.

Ref: https://datatracker.ietf.org/doc/html/rfc6902/#page-6

Reply:
my theory is that  if vm has `/spec/template/spec` without specific values for the rest of the hirarchy then the rest of the hierarchy gets initialized to default zero values and for pointer fields to null which later being replaced to a valid cpu object

for example:
spec:
  template:
    spec:
      architecture: amd64
      domain:

No cpu and no memory fields exist

it is equivalent to 
spec:
  template:
    spec:
      architecture: amd64
      domain:
        memory: null
        cpu: null
        
The replace operation will replace the null with the relevant object

example for failed operation will be trying to replace just the sockets value while it's parent the cpu doesn't exist (or is null) will fail as you guys suggesting

Reply:
You assume an existing implementation to be stable.
Defaults may change by being optimized or moved such that they will be executed at a different time.

You should not assume anything and try to make sure it always works.
If there is a need to override or add a key with the value, you better use `add` instead. `replace` is used when you need to assure the prev key already existed, which is not required here IIUC.

To be clear, there is still an assumption here that the `domain` key exists. If it does not exist, it will also fail. But I think the `domain` must be set, otherwise nothing will work anyway.

Reply:
you have a good point, i will adjust the code to perform specific add or replace operations were needed